openvdb/points/PointScatter.h

```diff
@@ -0,0 +1,676 @@
+//
+// *     Neither the name of DreamWorks Animation nor the names of
+// its contributors may be used to endorse or promote products derived
+// from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// IN NO EVENT SHALL THE COPYRIGHT HOLDERS' AND CONTRIBUTORS' AGGREGATE
+// LIABILITY FOR ALL CLAIMS REGARDLESS OF THEIR BASIS EXCEED US$250.00.
+//
+///////////////////////////////////////////////////////////////////////////
+//
+/// @author Nick Avramoussis
+///
+/// @file PointScatter.h
+///
+/// @brief Various point scattering methods for generating VDB Points.
+///
+///  Note that this architecture is heavily based off the similar
+///  functionality for more generic position scattering in
+///  tools/PointScatter.h, with an additional set of free functions for
+///  convenience.
+///
+///  All random number calls are made to the same generator to produce
+///  temporarily consistent results in relation to the provided seed. This
+///  comes with some multi-threaded performance trade-offs.
+///
+#ifndef OPENVDB_POINTS_POINT_SCATTER_HAS_BEEN_INCLUDED
+#define OPENVDB_POINTS_POINT_SCATTER_HAS_BEEN_INCLUDED
+
+#include <algorithm>
+#include <thread>
+
+#include <openvdb/openvdb.h>
+#include <openvdb/tree/LeafManager.h>
+#include <openvdb/tools/Prune.h>
+#include <openvdb/util/NullInterrupter.h>
+
+#include "PointDataGrid.h"
+#include "AttributeArray.h"
+#include "PointCount.h"
+
+#include <tbb/parallel_sort.h>
+#include <tbb/parallel_for.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace points {
+
+/// @brief Uniformly scatter a total amount of points in active regions
+///
+/// @param grid   A source grid. The resulting PointDataGrid will copy this grids
+///               transform and scatter in its active voxelized topology. Topology
+///               will be deactivated in areas where zero points are scattered.
+/// @param seed   A seed for the RandGenT
+/// @param count  The total number of points to scatter
+/// @note returns the scattered PointDataGrid
+template<
+    typename RandGenT,
+    typename GridT,
+    typename PointDataGridT = points::PointDataGrid>
```

Comment:
Sounds good - As GridT will most likely always be inferred that seems fine to go last, but I would have thought changing the generator type would be more common and should go first?

Reply:
Yes agreed, the ideal would be for RandGenT to have a default, but *also* be the first argument. If we shift the position of GridT, this won't compile because GridT doesn't have a default, but perhaps the simplest solution is to just give GridT a FloatGrid default as well? It'll typically always be inferred, but at least we can put RandGenT first and give it a default.